### ISSUE TITLE
Ability to validate local migrations without database

### ISSUE THREAD

**Issue by mattbroekhuis** [flyway/flyway#986]
Is there a way to run a sanity check on the local migration file(s)? Not comparing it against any actual database, but just a way to validate syntax (order, sequence numbers, file name, other things I'm not thinking of, etc)

**Comment by axelfontaine** [comment]
Thanks for your suggestion. Can you describe your scenario in a bit more detail? What is the real-world situation where you need this?

Cheers
Axel

**Comment by frederikd** [comment]
@axelfontaine I know this is an older issue, but we are also interested in this, so maybe a few words from our side on it:
We use feature branches and each feature branch can introduce new scripts. This can lead to conflicts on merge as two branches might both try to introduce the same script/version number. E.g. master is on version 1.2.3, and now two branches both want to introduce script 1.2.4, but with different file names for the different features.
This will lead to failed Flyway migrations as we have conflicting files for the same version number.

We currently find this only when we actually try to deploy the merged version as part of our CI pipeline, but it would be nice to be able to see this immediately as part of unit tests or other CI steps without having to actually execute the scripts or even have a DB. Then we could execute this sanity check on each build which would make the feedback loop in our CI/CD environment much quicker, and devs could see the issue in their local environment immediately without having to manually check script names or execute a local migration.

We already validate it all and it works okay, but it currently is quite late in the pipeline, and it would be nice to have this earlier in the feedback loop.

**Comment by janostik** [comment]
Facing same issue as @frederikd . Any progress on this?

**Comment by wgorski** [comment]
This is a common problem for all projects I've worked on where the number of developers is > 1. 
As stated before, feature branches might get out of sync which causes the application start to fail. Ideally we would like to discover such errors when building the project.
Currently, we've implemented our own validation script. It would be good to have a solution provided by flyway.

**Comment by esauser** [comment]
@wgorski can you share said validation script?

**Comment by wgorski** [comment]
We run a groovy script using gmavenplus-plugin

`def directory = "${project.basedir}/src/main/resources/schema/"

def files = []
new File(directory).traverse(type: groovy.io.FileType.FILES, nameFilter: ~/.*\.sql/) {it ->
  //check if repeatable migrations
  if (it.name =~ /R__.*$/) {
    def group = it.name =~ /(R__\d\d\d).*/
    files += group[0][1]
  } else {
    //for all none-repeatable migration
    files += it.name.replaceFirst(/__.*$/, "")
  }
}
if (files.size() != files.unique(false).size()) {
  log.error("Found duplicated migration versions:")
  files.groupBy().findAll {it.value.size() > 1}.each {
    log.error("  ${it.key}")
  }
  exit 1
}
`

**Comment by abbasadel** [comment]
Any progress on this guys. I also face this problem like many others. 

Thanks.

**Comment by vadim-hleif** [comment]
have the same issue 

it's so strange implement something by the hand before starting using some tool
e.g. liquibase works properly out of the box due to master file

**Comment by MikielAgutu** [comment]
@wgorski Can you explain what this script does beyond Flyways' [validate command](https://flywaydb.org/documentation/command/validate)? For instance, with verisoned migrations `V1__first.sql`, `V1__duplicate.sql`, running `flyway validate` will produce:

```
ERROR: Found more than one migration with version 1
Offenders:
-> C:\github\flyway-master\flyway-cli-sample\sql\V1__duplicate.sql (SQL)
-> C:\github\flyway-master\flyway-cli-sample\sql\V1__first.sql (SQL)
```

  There is also the new `validateMigrationNaming` option also (search for it [on this page](https://flywaydb.org/documentation/configfiles)).

@vadim-hleif Can you explain in more detail what sort of validation you would like to see in Flyway?

**Comment by vadim-hleif** [comment]
@MikielAgutu 
Hmm, this command uses real database for checking. 

But even if plugin will provide task for checking without real database, it also be inconvenient.

The problem is running this functional as merge check before all merges (e.g. in bitbucket merge checks run every time when you reload page).The merging without such check can produce errors on application startup, it's too late, it should be detected before merging (e.g. with master file in liquibase check works every time when you try to merge pr - if someone add new lines we will see conflict).

@wgorski's script solve does what i described. But it's so strange - copy script in every project, also in future versions of flyway script can require changes. Also it's groovy script, i know that bitbucket supports this, but i don't know how use it in github. Maybe it's real to do integration with github/bitbucket/etc.  (i think validate plugin command can't be used in such case. Agents should have maven/gradle and running the command will be so slowly)

**Comment by juliahayward** [comment]
This may be of interest? https://flywaydb.org/blog/using-git-hooks-with-flyway
